src/OpenTelemetry.AutoInstrumentation.Native/regex_utils.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#include "regex_utils.h"
+
+#include <regex>
+#include <sstream>
+
+// Include string utilities for conversion functions - after regex to avoid conflicts
+#include "string_utils.h"
+
+namespace trace
+{
+
+bool ExtractVersion(
+    const WSTRING& str, unsigned short& major, unsigned short& minor, unsigned short& build, unsigned short& revision)
+{
+    if (str.empty())
+    {
+        return false;
+    }
+
+#ifdef _WIN32
+    static std::wregex re(WStr("Version=([0-9]+)\\.([0-9]+)\\.([0-9]+)\\.([0-9]+)"));
+    std::wsmatch       match;
+
+    if (std::regex_search(str, match, re) && match.size() == 5)
+    {
+        WSTRINGSTREAM ss1(match.str(1));
+        ss1 >> major;
+        WSTRINGSTREAM ss2(match.str(2));
+        ss2 >> minor;
+        WSTRINGSTREAM ss3(match.str(3));
+        ss3 >> build;
+        WSTRINGSTREAM ss4(match.str(4));
+        ss4 >> revision;
+        return true;
+    }
+#else
+    // On Linux/MacOS, convert to narrow string for regex
+    std::string       narrow_str = ToString(str);
+    static std::regex re("Version=([0-9]+)\\.([0-9]+)\\.([0-9]+)\\.([0-9]+)");
+    std::smatch       match;
+
+    if (std::regex_search(narrow_str, match, re) && match.size() == 5)
+    {
+        major    = (unsigned short)std::stoi(match.str(1));
+        minor    = (unsigned short)std::stoi(match.str(2));
+        build    = (unsigned short)std::stoi(match.str(3));
+        revision = (unsigned short)std::stoi(match.str(4));
+        return true;
+    }
+#endif
+
+    return false;
+}
+
+WSTRING ExtractCulture(const WSTRING& str)
+{
+    if (str.empty())
+    {
+        return WSTRING();
+    }
+
+#ifdef _WIN32
+    static std::wregex re(WStr("Culture=([a-zA-Z0-9]+)"));
+    std::wsmatch       match;
+
+    if (std::regex_search(str, match, re) && match.size() == 2)
+    {
+        return match.str(1);
+    }
+#else
+    // On Linux/MacOS, convert to narrow string for regex
+    std::string       narrow_str = ToString(str);
+    static std::regex re("Culture=([a-zA-Z0-9]+)");
+    std::smatch       match;
+
+    if (std::regex_search(narrow_str, match, re) && match.size() == 2)
+    {
+        return ToWSTRING(match.str(1));
+    }
+#endif
+
+    return WSTRING();
+}
+
+bool ExtractPublicKeyToken(const WSTRING& str, unsigned char* data)
+{
+    if (str.empty() || data == nullptr)
+    {
+        return false;
+    }
+
+#ifdef _WIN32
+    static std::wregex re(WStr("PublicKeyToken=([a-fA-F0-9]{16})"));
+    std::wsmatch       match;
+
+    if (std::regex_search(str, match, re) && match.size() == 2)
+    {
+        for (int i = 0; i < 8; i++)
```

Comment:
The length of the array pointed to by `data` is now defined in two locations: The declaration of the array being passed to this function, and in the this loop. Should we add the length of `data` as a parameter to this method, or perhaps define it in a constant or `#define` value so that it can be shared by the two locations?

Reply:
Fixed in ee27e33deaa779e25a745d49a2d66a59571a839c
Passed the length as a parameter to this method, as I do not want to put here reference to instrumentation.h header.
It is reusing already existing const.